Added support for checking optional keyword arguments in quantity_input

adrn opened the pull request:
I have a function with complex behavior for default input, but the non-default values should always have a certain unit. For (a contrived) example:

```python
@u.quantity_input(t=u.day)
def shift_time(t, offset=None):
    if offset is None:    
        offset = np.median(t)
    else:
        # I would need to explicitly check for units of offset!
    return t - offset
```

On `master`, I have to check the units of the non-default `offset`. This PR adds support for validating the units of optional kwargs:

```python
@u.quantity_input(t=u.day, offset=(u.day,None))
def shift_time(t, offset=None):
    if offset is None:    
        offset = np.median(t)
    return t - offset
```

The way this is implemented in this PR, the arguments to `quantity_input` can now be iterable so that you have to explicitly say that you want to interpret `None` as the default value. Meaning, this doesn't work:

```python
@u.quantity_input(t=u.day, offset=u.day)
def shift_time(t, offset=None):
    # ...
```

If we want to support the above instead, I'm happy to implement that. 

A side effect of the way I implemented this is that now you can pass in multiple units to be validated against. In my own code, I sometimes have cases where I want a function to accept a length-like, velocity-like, or dimensionless Quantity. This would also enable that, e.g.:

```python
@u.quantity_input(xv=(u.pc, u.pc/u.Myr, u.dimensionless_unscaled))
def do_the_dynamics(xv):
    # ...
```

cc @mhvk

mhvk commented:
Am on holidays so may not get to looking at the actual code all that soon, but thought I should state here that I like the idea of the iterable. I'm not completely sure that any defaults given in the function defintion ever need to be checked, but since you'd probably also want to allow calling code to pass in (or *on*) any `offset=None`, that is somewhat immaterial.  Anyway, +1 to the general idea.

Cadair commented:
I like the idea of iterable but I would also just like it to ignore `None` i.e. I would rather have both features!
i.e this should work:
```python
@u.quantity_input(xv=(u.pc, u.pc/u.Myr, u.dimensionless_unscaled))
def do_the_dynamics(xv=None):
    # ...
```

Cadair commented:
I think you should be able to use inspect to get the default value for kwargs to implement the check that the arg is `None` and the default value is `None`.

Also this needs python 3 tests for the function annotation syntax.

adrn commented:
OK, I rebased and updated to support what is suggested above (auto-ignoring arguments with a default value of `None`). I'll have to rebase again once #3847 is merged because that will break some of this...

Cadair commented on astropy/units/tests/py3_test_quantity_annotations.py:
```diff
@@ -239,3 +239,27 @@ def myfunc_args(solarx: u.arcsec) -> u.deg:
     """
+
+@py3only
+def test_args_None():
```
Can you test multiple units in the annotations as well?

Cadair commented on astropy/units/decorators.py:
```diff
@@ -115,26 +117,51 @@ def wrapper(*func_args, **func_kwargs):
-                              "You may want to pass in an astropy Quantity instead."
-                                 .format(param.name, wrapped_function.__name__, error_msg))
+                    # short circuit if optional argument
+                    if arg is None:
```
This is confusing me a little, are you sure the following code would not trigger this condition:

```python
@u.quantity_input
def test(x: u.arcsec):
    print(x)

test(None)
```

adrn replied:
I guess I misunderstood your or @mhvk's comment above -- I thought you were both saying that you'd like to pass through a `None`?

Cadair replied:
I would say that we should allow `None` only if it's the default value of a keyword argument.

mhvk replied:
Yes, I think that was the idea: that it is useful for `None` to indicate a default, and that in that case it should not be checked for units. I think the generalization of this is that the default for a keyword should just always be passed on.

Cadair commented on astropy/units/decorators.py:
```diff
@@ -115,26 +117,51 @@ def wrapper(*func_args, **func_kwargs):
+
+                            try:
+                                equivalent = arg.unit.is_equivalent(target_unit,
+                                                                equivalencies=self.equivalencies)
```
indentation mis-match.

Cadair commented:
Oh, this also needs docs ;)

adrn commented:
Still needs: changelog entry, and notes in documentation

adrn commented:
@mhvk @Cadair Alright, I added some examples to the documentation and added an entry to the changelog. Since you last looked, I also cleaned up some of the code in the decorator itself by splitting it out into separate (private) functions.

Cadair commented on astropy/units/decorators.py:
```diff
@@ -5,10 +5,85 @@
+                  "You may want to pass in an astropy Quantity instead."
+                     .format(param_name, func_name, error_msg))
+
+    else:
```
Either this else statement or the above for statement has the incorrect indentation?

adrn replied:
@Cadair that belongs to a `for...else` statement -- see the for loop at L57

Cadair replied:
I did not know that was a thing, that's weird but awseome.

mhvk commented on astropy/units/decorators.py:
```diff
@@ -100,62 +175,42 @@ def wrapper(*func_args, **func_kwargs):
+
+                # If the argument value is None, and the default value is None,
+                #   pass through the None even if there is a target unit
+                if arg is None and param.default is None:
```
Could we generalize this slightly as `if arg is param.default` - I think the default never has to be checked. (Or is the case where no argument is given but a default is set already covered otherwise?)

Cadair replied:
I would like the default to be checked if it's a quantity. Seems to make logical sense to check for that case.

adrn replied:
I agree with @Cadair - I think the default should be checked, and the case where no argument is given but a default is set does get checked naturally because it will be passed as `arg` in the later code blocks.

mhvk replied:
I guess to me it makes more sense to check those defaults at function-definition time, but I think that can always be added later. So, for now, just leave as is, then.

mhvk commented on astropy/units/decorators.py:
```diff
@@ -100,62 +175,42 @@ def wrapper(*func_args, **func_kwargs):
+                #   specified
+                if isinstance(targets, str) or not isiterable(targets):
+                    targets = [targets]
+
```
I would suggest dealing with `None` here, so that the rest of the code can assume pure units:
```
elif None in targets:
    if arg is None:
        continue
    else:
        targets.remove(None)
```
(the `elif` will ensure `targets = None` fails -- which I think it should).

mhvk commented on astropy/units/decorators.py:
```diff
@@ -5,10 +5,85 @@
+    allowed_units = _get_allowed_units(targets)
+
+    # short-circuit if None is passed, and None is allowed
+    if arg is None and None in targets:
```
With my comment below, this can be removed -- I think it is somewhat illogical to deal with `None` here.

mhvk commented on astropy/units/decorators.py:
```diff
@@ -5,10 +5,85 @@
+
+    allowed_units = []
+    for target in targets:
+        if isinstance(target, str): # Could be a string unit or physical type
```
With my comment below, where I suggest dealing with `None` in the main code, we can insist all targets are either `Unit` or `string, and this `isinstance` check can be removed (since `Unit` instances pass unchanged through the `Unit` initlializer -- and this is very fast).

mhvk commented on astropy/units/tests/test_quantity_decorator.py:
```diff
@@ -246,4 +248,46 @@ def myfunc_args(x, y=10*u.deg):
+    y_target = u.km
+    y_unit = u.kpc
+
+    @u.quantity_input(x=[x_target, None], y=[y_target, None])
```
Put `None` first in one of the two, just as an increased sanity check.

adrn commented:
@mhvk Implemented most of your suggested changes. Anything else from you or @Cadair ?

mhvk commented on astropy/units/decorators.py:
```diff
@@ -100,62 +167,55 @@ def wrapper(*func_args, **func_kwargs):
+                # Check for None in the supplied list of allowed units and, if
+                #   present and the passed value is also None, ignore.
+                elif None in targets:
+                    valid_targets = [t for t in targets] # make a shallow copy
```
Good point about the copy -- otherwise we change the targets. But maybe just
```
if arg is None:
    continue
else:
    targets = [t for t in targets if t is not None
```

mhvk commented:
Change looks good!

adrn commented:
Cool - waiting for tests to complete, even if it's overkill! Feel free to merge if you notice the tests finish, otherwise I'll merge in a bit. Thanks for the feedback!
